tstest/integration.newTestEnv is flaky

https://github.com/tailscale/corp/actions/runs/3989076415/jobs/6841058944

<details>
  <summary>Click to expand test logs</summary>

```
panic: test timed out after 10m0s

goroutine 511 [running]:
testing.(*M).startAlarm.func1()
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:2036 +0xb4
created by time.goFunc
	/home/ubuntu/.cache/tailscale-go/src/time/sleep.go:176 +0x48

goroutine 1 [chan receive, 9 minutes]:
testing.tRunner.func1()
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1412 +0x5dc
testing.tRunner(0xc000228820, 0xc000161aa8)
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1452 +0x1bc
testing.runTests(0xc000224b40?, {0x11b06a0, 0xd, 0xd}, {0x4?, 0xa?, 0x11bbf40?})
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1844 +0x6d0
testing.(*M).Run(0xc000224b40)
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1726 +0x880
tailscale.com/tstest/integration.TestMain(0x0?)
	/home/ubuntu/go/pkg/mod/tailscale.com@v1.1.1-0.20230122221356-64547b2b86f3/tstest/integration/integration_test.go:57 +0xd4
main.main()
	_testmain.go:73 +0x308

goroutine 28 [syscall, 9 minutes]:
syscall.Syscall6(0xc00070f8d8?, 0xab958?, 0xc00070f8e8?, 0xacc18?, 0xc00070f918?, 0xb77c0?, 0xc00070f938?)
	/home/ubuntu/.cache/tailscale-go/src/syscall/syscall_linux.go:90 +0x34
os.(*Process).blockUntilWaitable(0xc000040540)
	/home/ubuntu/.cache/tailscale-go/src/os/wait_waitid.go:32 +0x7c
os.(*Process).wait(0xc000040540)
	/home/ubuntu/.cache/tailscale-go/src/os/exec_unix.go:22 +0x48
os.(*Process).Wait(...)
	/home/ubuntu/.cache/tailscale-go/src/os/exec.go:132
os/exec.(*Cmd).Wait(0xc000444420)
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:599 +0x78
os/exec.(*Cmd).Run(0x1?)
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:437 +0x58
os/exec.(*Cmd).CombinedOutput(0xc000444420)
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:707 +0x218
tailscale.com/tstest/integration.(*testNode).MustUp(0xc00035c230, {0x0, 0x0, 0x0?})
	/home/ubuntu/go/pkg/mod/tailscale.com@v1.1.1-0.20230122221356-64547b2b86f3/tstest/integration/integration_test.go:844 +0x25c
tailscale.com/tstest/integration.TestC2NPingRequest(0xc000288820)
	/home/ubuntu/go/pkg/mod/tailscale.com@v1.1.1-0.20230122221356-64547b2b86f3/tstest/integration/integration_test.go:387 +0x94
testing.tRunner(0xc000288820, 0xb81090)
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1446 +0x18c
created by testing.(*T).Run
	/home/ubuntu/.cache/tailscale-go/src/testing/testing.go:1493 +0x568

goroutine 34 [IO wait, 10 minutes]:
net.(*TCPListener).accept(0xc00033afa8)
	/home/ubuntu/.cache/tailscale-go/src/net/tcpsock_posix.go:142 +0x3c
net.(*TCPListener).Accept(0xc00033afa8)
	/home/ubuntu/.cache/tailscale-go/src/net/tcpsock.go:288 +0x68
net/http.(*Server).Serve(0xc00035e0f0, {0xccaee0, 0xc00033afa8})
	/home/ubuntu/.cache/tailscale-go/src/net/http/server.go:3070 +0x440
net/http/httptest.(*Server).goServe.func1()
	/home/ubuntu/.cache/tailscale-go/src/net/http/httptest/server.go:310 +0xa4
created by net/http/httptest.(*Server).goServe
	/home/ubuntu/.cache/tailscale-go/src/net/http/httptest/server.go:308 +0x9c

goroutine 36 [IO wait, 9 minutes]:
internal/poll.runtime_pollWait(0xffff9066c840, 0x72)
	/home/ubuntu/.cache/tailscale-go/src/runtime/netpoll.go:305 +0xa0
internal/poll.(*pollDesc).wait(0xc000092378, 0xc00031a000?, 0x1)
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_poll_runtime.go:84 +0xc0
internal/poll.(*pollDesc).waitRead(...)
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc000092360, {0xc00031a000, 0x8000, 0x8000})
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_unix.go:167 +0x2f8
os.(*File).read(...)
	/home/ubuntu/.cache/tailscale-go/src/os/file_posix.go:31
os.(*File).Read(0xc000308068, {0xc00031a000, 0x8000, 0x8000})
	/home/ubuntu/.cache/tailscale-go/src/os/file.go:119 +0x9c
io.copyBuffer({0xcc30c0, 0xc0003806f0}, {0xcc2d40, 0xc000308068}, {0x0, 0x0, 0x0})
	/home/ubuntu/.cache/tailscale-go/src/io/io.go:427 +0x1f0
io.Copy(...)
	/home/ubuntu/.cache/tailscale-go/src/io/io.go:386
os/exec.(*Cmd).writerDescriptor.func1()
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:407 +0x60
os/exec.(*Cmd).Start.func1(0xc00028e040)
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:544 +0x38
created by os/exec.(*Cmd).Start
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:543 +0x988

goroutine 73 [IO wait, 9 minutes]:
internal/poll.runtime_pollWait(0xffff8879f3d8, 0x72)
	/home/ubuntu/.cache/tailscale-go/src/runtime/netpoll.go:305 +0xa0
internal/poll.(*pollDesc).wait(0xc00066e5b8, 0xc000394600?, 0x1)
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_poll_runtime.go:84 +0xc0
internal/poll.(*pollDesc).waitRead(...)
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_poll_runtime.go:89
internal/poll.(*FD).Read(0xc00066e5a0, {0xc000394600, 0x200, 0x200})
	/home/ubuntu/.cache/tailscale-go/src/internal/poll/fd_unix.go:167 +0x2f8
os.(*File).read(...)
	/home/ubuntu/.cache/tailscale-go/src/os/file_posix.go:31
os.(*File).Read(0xc0003082f8, {0xc000394600, 0x200, 0x200})
	/home/ubuntu/.cache/tailscale-go/src/os/file.go:119 +0x9c
bytes.(*Buffer).ReadFrom(0xc00037a3f0, {0xcc2d40, 0xc0003082f8})
	/home/ubuntu/.cache/tailscale-go/src/bytes/buffer.go:202 +0xec
io.copyBuffer({0xcc1060, 0xc00037a3f0}, {0xcc2d40, 0xc0003082f8}, {0x0, 0x0, 0x0})
	/home/ubuntu/.cache/tailscale-go/src/io/io.go:413 +0x[158](https://github.com/tailscale/corp/actions/runs/3989076415/jobs/6841058944#step:5:159)
io.Copy(...)
	/home/ubuntu/.cache/tailscale-go/src/io/io.go:386
os/exec.(*Cmd).writerDescriptor.func1()
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:407 +0x60
os/exec.(*Cmd).Start.func1(0xc00028e5e0)
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:544 +0x38
created by os/exec.(*Cmd).Start
	/home/ubuntu/.cache/tailscale-go/src/os/exec/exec.go:543 +0x988
FAIL	tailscale.com/tstest/integration	600.202s
```

</details>

--- a/cmd/tailscale/cli/up.go
+++ b/cmd/tailscale/cli/up.go
@@ -496,11 +496,23 @@ func runUp(ctx context.Context, cmd string, args []string, upArgs upArgsT) (retE
 	running := make(chan bool, 1) // gets value once in state ipn.Running
 	pumpErr := make(chan error, 1)
 
-	var printed bool // whether we've yet printed anything to stdout or stderr
-	var loginOnce sync.Once
-	startLoginInteractive := func() { loginOnce.Do(func() { localClient.StartLoginInteractive(ctx) }) }
+	// localAPIMu should be held while doing mutable LocalAPI calls
+	// to the backend. In particular, it prevents StartLoginInteractive from
+	// being called from the watcher goroutine while the Start call from
+	// the other goroutine is in progress.
+	// See https://github.com/tailscale/tailscale/issues/7036#issuecomment-2053771466
+	// TODO(bradfitz): simplify this once #11649 is cleaned up and Start is
+	// hopefully removed.
+	var localAPIMu sync.Mutex
+
+	startLoginInteractive := sync.OnceFunc(func() {
+		localAPIMu.Lock()
+		defer localAPIMu.Unlock()
+		localClient.StartLoginInteractive(ctx)
+	})
 
 	go func() {
+		var printed bool // whether we've yet printed anything to stdout or stderr
 		for {
 			n, err := watcher.Next()
 			if err != nil {
@@ -574,12 +586,14 @@ func runUp(ctx context.Context, cmd string, args []string, upArgs upArgsT) (retE
 	// Special case: bare "tailscale up" means to just start
 	// running, if there's ever been a login.
 	if simpleUp {
+		localAPIMu.Lock()
 		_, err := localClient.EditPrefs(ctx, &ipn.MaskedPrefs{
 			Prefs: ipn.Prefs{
 				WantRunning: true,
 			},
 			WantRunningSet: true,
 		})
+		localAPIMu.Unlock()
 		if err != nil {
 			return err
 		}
@@ -596,10 +610,13 @@ func runUp(ctx context.Context, cmd string, args []string, upArgs upArgsT) (retE
 		if err != nil {
 			return err
 		}
-		if err := localClient.Start(ctx, ipn.Options{
+		localAPIMu.Lock()
+		err = localClient.Start(ctx, ipn.Options{
 			AuthKey:     authKey,
 			UpdatePrefs: prefs,
-		}); err != nil {
+		})
+		localAPIMu.Unlock()
+		if err != nil {
 			return err
 		}
 		if upArgs.forceReauth {

--- a/tstest/integration/integration_test.go
+++ b/tstest/integration/integration_test.go
@@ -345,7 +345,6 @@ func TestConfigFileAuthKey(t *testing.T) {
 
 func TestTwoNodes(t *testing.T) {
 	tstest.Shard(t)
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/3598")
 	tstest.Parallel(t)
 	env := newTestEnv(t)
 
@@ -400,7 +399,6 @@ func TestTwoNodes(t *testing.T) {
 // PeersRemoved set) saying that the second node disappeared.
 func TestIncrementalMapUpdatePeersRemoved(t *testing.T) {
 	tstest.Shard(t)
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/3598")
 	tstest.Parallel(t)
 	env := newTestEnv(t)
 
@@ -658,17 +656,14 @@ func TestNoControlConnWhenDown(t *testing.T) {
 	d2 := n1.StartDaemon()
 	n1.AwaitResponding()
 
-	st := n1.MustStatus()
-	if got, want := st.BackendState, "Stopped"; got != want {
-		t.Fatalf("after restart, state = %q; want %q", got, want)
-	}
+	n1.AwaitBackendState("Stopped")
 
 	ip2 := n1.AwaitIP4()
 	if ip1 != ip2 {
 		t.Errorf("IPs different: %q vs %q", ip1, ip2)
 	}
 
-	// The real test: verify our daemon doesn't have an HTTP request open.:
+	// The real test: verify our daemon doesn't have an HTTP request open.
 	if n := env.Control.InServeMap(); n != 0 {
 		t.Errorf("in serve map = %d; want 0", n)
 	}
@@ -1007,7 +1002,6 @@ func newTestEnv(t testing.TB, opts ...testEnvOpt) *testEnv {
 	if runtime.GOOS == "windows" {
 		t.Skip("not tested/working on Windows yet")
 	}
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/7036")
 	derpMap := RunDERPAndSTUN(t, logger.Discard, "127.0.0.1")
 	logc := new(LogCatcher)
 	control := &testcontrol.Server{
@@ -1389,6 +1383,10 @@ func (n *testNode) AwaitIP6() netip.Addr {
 
 // AwaitRunning waits for n to reach the IPN state "Running".
 func (n *testNode) AwaitRunning() {
+	n.AwaitBackendState("Running")
+}
+
+func (n *testNode) AwaitBackendState(state string) {
 	t := n.env.t
 	t.Helper()
 	if err := tstest.WaitFor(20*time.Second, func() error {
@@ -1396,8 +1394,8 @@ func (n *testNode) AwaitRunning() {
 		if err != nil {
 			return err
 		}
-		if st.BackendState != "Running" {
-			return fmt.Errorf("in state %q", st.BackendState)
+		if st.BackendState != state {
+			return fmt.Errorf("in state %q; want %q", st.BackendState, state)
 		}
 		return nil
 	}); err != nil {

--- a/wgengine/magicsock/magicsock_test.go
+++ b/wgengine/magicsock/magicsock_test.go
@@ -626,7 +626,7 @@ func (localhostListener) ListenPacket(ctx context.Context, network, address stri
 }
 
 func TestTwoDevicePing(t *testing.T) {
-	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/1277")
+	flakytest.Mark(t, "https://github.com/tailscale/tailscale/issues/11762")
 	l, ip := localhostListener{}, netaddr.IPv4(127, 0, 0, 1)
 	n := &devices{
 		m1:     l,

